src/lsm/lsm_cursor.c

```diff
@@ -889,6 +889,79 @@ err:	__clsm_leave(clsm);
 }
 
 /*
+ * __clsm_next_random --
+ *	WT_CURSOR->next method for the LSM cursor type when configured with
+ * next_random.
+ */
+static int
+__clsm_next_random(WT_CURSOR *cursor)
+{
+	WT_CURSOR_LSM *clsm;
+	WT_CURSOR *c;
+	WT_DECL_RET;
+	WT_LSM_CHUNK *chunk;
+	WT_SESSION_IMPL *session;
+	uint64_t checked_docs, i, rand_doc, total_docs;
+	const char *cfg[3];
+
+	clsm = (WT_CURSOR_LSM *)cursor;
+
+	CURSOR_API_CALL(cursor, session, next, NULL);
+	WT_CURSOR_NOVALUE(cursor);
+	WT_ERR(__clsm_enter(clsm, false, false));
+
+	cfg[0] = WT_CONFIG_BASE(session, WT_SESSION_drop);
+	cfg[1] = "next_random=true";
+	cfg[2] = NULL;
+
+	/* Initialize to the 0th chunk */
+	chunk = clsm->lsm_tree->chunk[0];
+
+	/* We need to know roughly how many docs are in the tree */
+retry:	for (total_docs = i = 0; i < clsm->lsm_tree->nchunks; i++) {
+		total_docs += clsm->lsm_tree->chunk[i]->count;
+	}
+
+	/*
+	 * We pick one of the docs at random from the tree to pick a random
+	 * chunk, weighted by the size of the chunks.
+	 */
+	rand_doc = __wt_random(&session->rnd) % total_docs;
+	for (checked_docs = i = 0; i < clsm->lsm_tree->nchunks; i++) {
+		checked_docs += clsm->lsm_tree->chunk[i]->count;
```

Comment:
Understood re: locks. I see two methods of approaching this problem:
Loop through twice under the lock
Loop through once, creating a static "copy" of the chunks array as it is "now" and then iterate over that

I'm not sure which would be faster in this case. I lean towards the former but am not sure how to judge this.